tools/toolchain/clang_lto_pgo.sh

```diff
@@ -0,0 +1,108 @@
+#!/bin/bash -uex
+
+if [ -z "${CLANG_BUILD}" ]; then
+	echo "Skip building optimized clang"
+	exit 0
+elif [ "${CLANG_BUILD}" = "SKIP" ]; then
+        echo "Skip building optimized clang"
+        exit 0
+elif [ "${CLANG_BUILD}" = "BUILD" ]; then
+	echo "build optimized clang"
+elif [ "${CLANG_BUILD}" = "INSTALL" ]; then
+	echo "build and install optimized clang"
+else
+	echo "Not sure what to do with ${CLANG_BUILD}"
+	exit 1
+fi
+# Which Scylla branch to train on
+SCYLLA_BRANCH=master
+
+# Which LLVM release to build
+LLVM_TAG=15.0.7
+
+if [ -d "/optimized_clang" ];
+then
+	DIR="/optimized_clang"
+else
+	DIR="${PWD}/optimized_clang"
+	mkdir -p "${DIR}"
+fi
+cd "${DIR}"
+
+# Download the scylla codebase for training.
+# Scylla's dependencies are already installed.
+git clone https://github.com/scylladb/scylla --branch ${SCYLLA_BRANCH} --depth=1
+git -C scylla submodule update --init --depth=1
+
+# Clone, patch and bootstrap the newest Clang and BOLT.
+git clone https://github.com/llvm/llvm-project --branch main --depth=1 stage-0
+cd stage-0
+
+cmake -B build -S llvm -DLLVM_ENABLE_PROJECTS='clang;lld;bolt' -DCMAKE_BUILD_TYPE=Release -DLLVM_TARGETS_TO_BUILD=X86 -G Ninja -DLLVM_ENABLE_RUNTIMES='compiler-rt' -DCOMPILER_RT_BUILD_SANITIZERS=Off -DCOMPILER_RT_DEFAULT_TARGET_ONLY=On -DLLVM_INCLUDE_BENCHMARKS=No -DLLVM_INCLUDE_EXAMPLES=No -DLLVM_INCLUDE_TESTS=No
+cmake --build build --
+
+USE_NEW_COMPILER=(-DCMAKE_C_COMPILER="$DIR/stage-0/build/bin/clang" -DCMAKE_CXX_COMPILER="$DIR/stage-0/build/bin/clang++" -DLLVM_USE_LINKER="$DIR/stage-0/build/bin/ld.lld")
+COMMON_OPTS=(-DLLVM_ENABLE_PROJECTS='clang' -DCMAKE_BUILD_TYPE=Release -DLLVM_TARGETS_TO_BUILD=X86 -G Ninja -DLLVM_ENABLE_LTO=Thin -DLLVM_BUILD_RUNTIME=No -DCLANG_DEFAULT_PIE_ON_LINUX=OFF -DLLVM_INCLUDE_BENCHMARKS=OFF -DLLVM_INCLUDE_EXAMPLES=OFF -DLLVM_INCLUDE_TESTS=OFF )
+
+# Build a PGO-optimized compiler using the boostrapped compiler.
+# Choose a version compatible with your system.
+# If you change it, remember to change clang-15 in the later part of the script to the appropriate name as well.
+git fetch --depth=1 origin tag llvmorg-${LLVM_TAG}
```

Comment:
There's the one that is compiling clang and there's the clang we wish to use. May or may not be the same.

Reply:
We can compile clang with the one provided by the distribution. Otherwise we have three clangs: one to compile clang, one to compile the clang we wish to use, and the clang we wish to use.

Reply:
https://github.com/llvm/llvm-project/issues/55004 is the reason we can't, AFAIK.

Reply:
Ah. Well, clang 16 is on  the way (rc2 now). So be Fedora 38 the packaged compiler will be okay.

We can wait for Fedora 38, or ask Fedora to backport, or do a two-stage build like in the patch.